core/modules/widgets/range.js

```diff
@@ -62,23 +64,102 @@ RangeWidget.prototype.render = function(parent,nextSibling) {
-			value = this.wiki.extractTiddlerDataItem(tiddler,this.tiddlerIndex,this.defaultValue || "");
+			value = this.wiki.extractTiddlerDataItem(tiddler,this.tiddlerIndex,this.defaultValue);
 		} else {
 			if($tw.utils.hop(tiddler.fields,fieldName)) {
 				value = tiddler.fields[fieldName] || "";
 			} else {
-				value = this.defaultValue || "";
+				value = this.defaultValue;
 			}
 		}
 	}
 	return value;
 };
 
+RangeWidget.prototype.prepareAttributes = function() {
+	var variables = Object.create(null);
+	// Trigger actions. Use variables = {key:value, key:value ...}
+	$tw.utils.each(this.attributes,function(val,key) {
+		if(key.charAt(0) !== "$") {
+			variables["attr-" + key] = val;
+		}
+	});
+	// "tiddler" parameter may be missing in the widget call. See .execute() below
+	// the "value" is needed.
+	variables = $tw.utils.extend(variables, {"actionValue": this.inputDomNode.value,
+			"attr-tiddler": this.tiddlerTitle,
+			"attr-field": this.tiddlerField
+/*			"attr-value": this.inputDomNode.value*/   });
+	return variables;
+}
+
+RangeWidget.prototype.handleMouseDownEvent = function(event) {
+	this.mouseDown = true; // TODO remove once IE is gone.
+
+console.log("mouse down",event);
+
+	this.handleEvent(event);
+	// Trigger actions
+	if(this.actionsMouseDown) {
+		var variables = this.prepareAttributes() // TODO this line will go into the function call below.
+		this.invokeActionString(this.actionsMouseDown,this,event,variables);
+	}
+	this.oldValue = this.inputDomNode.value; // TODO remove this line once IE is gone!
+}
+
+RangeWidget.prototype.handleMouseUpEvent = function(event) {
+	this.mouseDown = false; // TODO remove once IE is gone.
+
+console.log("mouse up",event);
+
+	this.handleEvent(event);
+	// Trigger actions
+	if(this.actionsMouseUp) {
+		var variables = this.prepareAttributes() // TODO this line will go into the function call below.
+		this.invokeActionString(this.actionsMouseUp,this,event,variables);
+	}
+	// TODO remove the following if() once IE is gone!
+	if ($tw.browser.isIE) {
+		if (this.oldValue !== this.inputDomNode.value) {
+			this.handleChangeEvent(event);
+			this.oldValue = this.inputDomNode.value;
+		}
+	}
+}
+
+RangeWidget.prototype.handleChangeEvent = function(event) {
+	if (this.mouseDown) {  // TODO refactor this function once IE is gone.
+		this.handleInputEvent(event);
+	} else {
+
+console.log("change",event);
+
+		this.handleEvent(event);
+		// Trigger actions
+		if(this.actionsChange) {
+			var variables = this.prepareAttributes() // TODO this line will go into the function call below.
+			this.invokeActionString(this.actionsChange,this,event,variables);
+		}
+	}
+};
+
 RangeWidget.prototype.handleInputEvent = function(event) {
+console.log("input",event);
+
+	this.handleEvent(event);
+	// Trigger actions
+	if(this.actionsInput) {
+		// "tiddler" parameter may be missing. See .execute() below
+		var variables = this.prepareAttributes() // TODO this line will go into the function call below.
```

Comment:
currently `handleEvent()` does debouncing and only calls `wiki.setText` if the value of the widget has changed. We should similarly debounce the actions, at least the ones for events like `input` that fire very frequently:

```
if(this.getValue() !== this.inputDomNode.value) {
    this.invokeActionString(this.actionsInput,this,event,variables);
}
```

Reply:
You are right. Isn't this the case for every widget?

Reply:
input or change(on IE) are only fired, if the value has been changed. ... IMO this is no debouncing. 

The line was introduced with commit e84c422e5091c02f55db4027faa9ba840e2aee6c ... But I don't see a difference, if I remove it. ... @Jermolene Do you remember the problem?

Reply:
@pmario according to the spec you should be right. I thought I ran into some issue with this last I worked on this widget. I can do some tests to confirm tomorrow.

Reply:
@pmario I can confirm that the debouncing does not seem necessary on the input event. In both Chrome and FF I cannot get the event to fire more than once per change of value.